File: backend/routes/cohorts.py
from flask import Blueprint, request, jsonify
from services.cohorts_services import CohortService
from ..model import User
from ..model import Cohort
from datetime import datetime

cohorts_routes = Blueprint('cohorts_routes', __name__)

# Create a new cohort (public or private)
@cohorts_routes.route('/create', methods=['POST'])
def create_cohort():
    data = request.get_json()
    
    # Validate input data
    if not data.get('name') or not data.get('admin_id') or not data.get('start_date'):
        return jsonify({"error": "Name, admin ID, and start date are required"}), 400

    # Ensure start_date is in proper format
    try:
        start_date = datetime.strptime(data['start_date'], '%Y-%m-%d %H:%M:%S')
    except ValueError:
        return jsonify({"error": "Invalid date format for start_date. Use YYYY-MM-DD HH:MM:SS"}), 400
    
    cohort_data = {
        'name': data['name'],
        'description': data.get('description', ''),
        'start_date': start_date,
        'end_date': data.get('end_date', None),
        'is_active': data.get('is_active', True),
        'admin_id': data['admin_id']
    }

    # Create cohort using the service
    cohort_id = CohortService.create_cohort(cohort_data)
    return jsonify({"message": "Cohort created successfully", "cohort_id": cohort_id}), 201

# List all cohorts
@cohorts_routes.route('/', methods=['GET'])
def list_cohorts():
    cohorts = CohortService.get_all_cohorts()
    return jsonify(cohorts), 200

# Update a cohort (e.g., change its name or description)
@cohorts_routes.route('/<cohort_id>', methods=['PUT'])
def update_cohort(cohort_id):
    data = request.get_json()

    # Validate input data
    if not data:
        return jsonify({"error": "No data provided for update"}), 400

    # Update the cohort using the service
    updated_cohort = CohortService.update_cohort(cohort_id, data)
    if not updated_cohort:
        return jsonify({"error": "Cohort not found"}), 404
    
    return jsonify({"message": "Cohort updated successfully"}), 200

# Add a user to a cohort
@cohorts_routes.route('/<cohort_id>/join', methods=['POST'])
def join_cohort(cohort_id):
    data = request.get_json()
    
    if 'user_id' not in data:
        return jsonify({"error": "User ID is required"}), 400

    user_id = data['user_id']
    
    # Join cohort using the service
    success = CohortService.add_user_to_cohort(cohort_id, user_id)
    if not success:
        return jsonify({"error": "Unable to join cohort"}), 400

    return jsonify({"message": "User successfully added to cohort"}), 200

File: backend/routes/cohorts.js
// backend/routes/cohorts.js
const express = require('express');
const router = express.Router();
const fs = require('fs');
const path = require('path');

const dataPath = path.join(__dirname, '../data/cohorts.json');

// Read cohorts data from file
const getCohortsData = () => {
  try {
    return JSON.parse(fs.readFileSync(dataPath));
  } catch (err) {
    return [];
  }
};

// Write cohorts data to file
const saveCohortsData = (data) => {
  fs.writeFileSync(dataPath, JSON.stringify(data, null, 2));
};

// Get all cohorts
router.get('/', (req, res) => {
  res.json(getCohortsData());
});

// Add a new cohort
router.post('/create', (req, res) => {
  const { name, adminId } = req.body;
  if (!name || !adminId) {
    return res.status(400).json({ error: 'Cohort name and adminId are required' });
  }

  const cohorts = getCohortsData();
  const newCohort = {
    id: Date.now(),
    name,
    adminId,  // adminId is passed when creating the cohort
    members: []
  };
  cohorts.push(newCohort);
  saveCohortsData(cohorts);
  res.status(201).json(newCohort);
});

// Delete a cohort
router.delete('/:id', (req, res) => {
  const cohorts = getCohortsData();
  const updatedCohorts = cohorts.filter(cohort => cohort.id !== parseInt(req.params.id));

  if (updatedCohorts.length === cohorts.length) {
    return res.status(404).json({ error: 'Cohort not found' });
  }

  saveCohortsData(updatedCohorts);
  res.sendStatus(204); // No content, successful deletion
});

// Add a member to a cohort
router.post('/:id/addMember', (req, res) => {
  const { name } = req.body;
  if (!name) {
    return res.status(400).json({ error: 'Member name is required' });
  }

  const cohorts = getCohortsData();
  const cohort = cohorts.find(cohort => cohort.id === parseInt(req.params.id));

  if (!cohort) {
    return res.status(404).json({ error: 'Cohort not found' });
  }

  cohort.members.push({ name, role: 'user' }); // Default role is 'user'
  saveCohortsData(cohorts);
  res.status(201).json(cohort);  // Return updated cohort
});

// Remove a member from a cohort (Only accessible to the admin)
router.post('/:id/removeMember', (req, res) => {
  const { adminId, memberName } = req.body;

  // Ensure both adminId and memberName are provided
  if (!adminId || !memberName) {
    return res.status(400).json({ error: 'adminId and memberName are required' });
  }

  const cohorts = getCohortsData();
  const cohort = cohorts.find(cohort => cohort.id === parseInt(req.params.id));

  if (!cohort) {
    return res.status(404).json({ error: 'Cohort not found' });
  }

  // Check if the adminId matches the cohort's adminId
  if (cohort.adminId !== adminId) {
    return res.status(403).json({ error: 'Only the admin can remove members from this cohort' });
  }

  // Find the member in the cohort's members list
  const memberIndex = cohort.members.findIndex(member => member.name === memberName);
  if (memberIndex === -1) {
    return res.status(404).json({ error: 'Member not found in this cohort' });
  }

  // Remove the member
  cohort.members.splice(memberIndex, 1);
  saveCohortsData(cohorts);

  res.json({ message: `Member ${memberName} has been removed from cohort ${cohort.name}` });
});

module.exports = router;
